### PR TITLE
APIv4 - Add Individual, Household, Organization pseudo-entities

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
This extends APIv4 with 3 pseudo-entities for the 3 base contact types.

Before
----------------------------------------
```
// Get individuals
Contact::get()->addWhere('contact_type', '=', 'Individual')...

// Create individual
Contact::create()->addValue('contact_type', 'Individual')...
```

After
----------------------------------------
The above still works of course, but now you can:
```
// Get individuals
Individual::get()...

// Create individual
Individual::create()...
```

Comments
-----------------
This aims to address a misalignment of what various subsystems consider to be an "Entity".

1. Custom fields can extend "entities" 'Contact', 'Individual', 'Organization', 'Household'...
2. Afforms can be for "entities" 'Individual', 'Organization', or 'Household' (but _not_ 'Contact')
3. API only has 'Contact' entity.

Because of this, mapping custom fields or Afform entities to API entities always takes some wrangling. And as we do more with Afforms, the amount of wrangling increases. I finally got tired of it and thought I'd circle back to this old idea. When @totten and I first discussed it a few years ago, it seemed too hard (how would you, for example, ensure this works across joins, or enforce permissions?) The structure of the APIv4 has improved since then so that it seemed doable (and this PR adds unit tests to ensure it *does* work across joins and enforce permissions).

Technical Details
----------------------------------------
The `Individual`, `Organization`, etc. APIs are exactly like the `Contact` API (they in fact *are* the `Contact` API) plus one predefined parameter you don't have to set (`contact_type`).

It's analogous to the difference between `civicrm_api()` and `civicrm_api3()`/`civicrm_api4()`. The former requires `$params['version']` and the latter insert it for you (it's always `3` or `4` respectively). Likewise, calling the `Contact` api requires a `contact_type` param, but calling the `Individual`, `Organization`, etc. APIs will insert that value for you.

This code uses `CRM_Contact_BAO_ContactType::basicTypes()` throughout (rather than a hard-coded list like `['Individual', 'Organization', 'Household']`. So you can disable contact types and the corresponding api will be disabled (or at least, non-funtional). In theory you could also *add* a base contact type (like `Robot`) and this would handle it fine (you'd need to add the file `Civi\Api4\Robot.php` but everything else should be automatic).

### CODE REVIEW

**Comment by totten** [comment]
> Afforms can be for 'Individual', 'Organization', or 'Household' (but not 'Contact')

Oh, nice. That's sounds like it'll be a significant simplification  (in the mental-model at minimum, and probably in the code too).

**Comment by mattwire** [comment]
A few questions:
- How does this work with contact subtypes?
- Is an upgrade path required for existing afforms?
- Many sites have "Household" disabled. Will this cause problems?

**Comment by colemanw** [comment]
> A few questions:
> 
> * How does this work with contact subtypes?

No differently than the Contact api. The 'Individual', 'Organization', etc. APIs are exactly like the 'Contact' api (they in fact *are* the Contact API) plus one predefined parameter you don't have to set (`contact_type`).

It's analogous to the difference between `civicrm_api()` and `civicrm_api3()`/`civicrm_api4()`. The former requires `$params['version']` and the latter inserts it for you (it's always `3` or `4` respectively). Likewise, calling the Contact api requires a `contact_type` param, but calling the 'Individual', 'Organization', etc. APIs will insert that value for you.

> * Is an upgrade path required for existing afforms?

No. I may followup with some refactoring of the Afform code to simplify things now that this PR allows 1-1 matching between afform entities and api entities, but it's all backward-compatible.

> * Many sites have "Household" disabled. Will this cause problems?

No, this code uses `CRM_Contact_BAO_ContactType::basicTypes()` throughout (rather than a hard-coded list like `['Individual', 'Organization', 'Household']` and that function returns only base contact types that are enabled. In theory you could also *add* a base contact type and this would handle it fine.

File: Civi/Api4/Contact.php [hunk 1]
```diff
@@ -96,8 +96,37 @@ public static function getDuplicates($checkPermissions = TRUE) {
...
+   */
+  public static function getInfo(): array {
+    $info = parent::getInfo();
+    $contactType = static::getEntityName();
+    // Adjust info for child classes (Individual, Organization, Household)
+    if ($contactType !== 'Contact') {
+      $info['icon'] = \CRM_Contact_BAO_ContactType::getContactType($contactType)['icon'] ?? $info['icon'];
+      $info['type'] = ['DAOEntity', 'ContactType'];
+      // This forces the value into get and create api actions
+      $info['where'] = ['contact_type' => $contactType];
```
Comment (colemanw) [inline Civi/Api4/Contact.php:119]: @totten this is the new mechanism that drives this whole thing. You can now define, at the api4-entity level, a `where` property which api4 will force into get, create and getFields. 

We could re-use this to, for example, create an `EventTemplate` api which extends `Event` but adds `$info['where'] = ['is_template' => TRUE];` and then that api would only ever get or create event templates.

Comment (totten) [reply Civi/Api4/Contact.php:119]: That's, pretty cool. It does help to give a general sense to the bigger PR. :+1:

**Comment by totten** [comment]
@colemanw I was playing around with some calls. Should one be concerned about this comparison:

```
# Try to read a Household using Organization API
$ cv api4 Organization.get +w id=20 
[]
```

That's ^^ good. Then compare:

```
## Try to modify a Household using Organization API
$ cv api4 Organization.update +w id=20 +v do_not_email=1
[
    {
        "id": 20,
        "contact_type": "Household",
        "display_name": "Cruz-Wattson family",
        "do_not_email": true,
        "do_not_phone": false,
```

The update seems to go through?

**Comment by colemanw** [comment]
@totten - actually that makes sense based on how APIv4 writes records. Since you passed it a single id, it bypasses `getBatchRecords` and just assumes you know what you're doing. But if you'd passed it any other criteria it would have run it through `getBatchRecords` and that would have gotten the contact_type injected by the fascade. IMO the performance benefit of being able to update a record by id is an acceptable trade-off.
I've added a unit test to demonstrate:
https://github.com/civicrm/civicrm-core/blob/d176ac0576d4cdefe1e0de648c81ccc9ca969cff/tests/phpunit/api/v4/Entity/ContactTypeTest.php#L171-L199

**Comment by totten** [comment]
> (@colemanw) IMO the performance benefit of being able to update a record by id is an acceptable trade-off.

OK, so the performance benefit feels more real/immediate, which is fair.

I should note - historically, in APIv3, after some years, there were derived features like [CiviProxy's `rest_allowed_actions`](https://docs.civicrm.org/civiproxy/en/latest/configuration/) and [CiviConnect's WhitelistRule](https://github.com/civicrm/civicrm-core/blob/6ef03256261c60a017a4feb23d95446771802037/Civi/API/WhitelistRule.php#L14) which treated the API entity-name as part of the security-guard. I don't think there's an analog in APIv4 world right now, but IMHO the pattern has some timeless appeal, and I wouldn't be surprised if it appears again. (Perhaps in JWT or an API-key admin system.) 

That's not today's problem -- we don't need to make that work. It's just context. For today...

I think we should communicate that the entity-aliasing mechanism as a convenience thing and ***not*** a security thing. (From a security POV, `Individual` API  === `Contact` API.) Which place is best for communicating that? Maybe docblocks of `Civi\Api4\{Individual,Organization,Household}`?

**Comment by colemanw** [comment]
@totten yea, this PR definitely makes answering the question "what entity is this?" more complicated. Anyone implementing an ApiWrapper hook will have a new gotcha to contend with.
But in other ways it's an improvement 🤷🏻 
FWIW to solve that problem internally added this method:
https://github.com/civicrm/civicrm-core/blob/d176ac0576d4cdefe1e0de648c81ccc9ca969cff/Civi/Api4/Utils/CoreUtil.php#L70-L75

**Comment by colemanw** [comment]
@totten I've pushed some docs ^

**Comment by totten** [comment]
Thanks @colemanw. Comments look good. This does have a lot of test coverage. Merging.

**Pull request merged**